**Goals**

1. Set up a pipeline to incorporate the imputation
2. Do a random forest regressor to identify important features
3. Do a test run with one model (linear, most likely) that computes:
    - MSE for predicting PCIAT-Total
    - MSE for predicting sii when computed from predicted PCIAT-Total
    - MSE for predicting sii directly
    - kappa for predicting sii when computed from predicted PCIAT-Total
    - kappa for predicting sii directly
4. After getting the model working, measure these things for out-of-the box:
    - multiple linear regression
    - knn regression
    - random forest
    - support vector
    - gradient boost
    - adaboost
    - xgboost
5. After identifying a promising out-of-the-box model, try tuning it
6. Try implementing a sequential predictor (either logistic regression or random forest) that:
    - Starts by predicting 3's vs. non-threes
    - Predicts 2's vs. non-twos from the remaining cases
    - etc.
7. Try using different models for doing this sequential prediction

In [1]:
import pandas as pd
import numpy as np

# CustomImputers includes two imputers (using KNN and interative imputing) for this dataset
# and also two functions to transform fitnessgram values into fitnessgram zones, and to round sii values
from CustomImputers import *

**Loading the Data**

For the purpose of developing our model(s), we'll work with data that include the imputed outcome (PCIAT_Total and/or sii) scores AND have cleaned predictors.

In the final version of our code, we'll work with data with cleaned predictors but won't have any access to the outcome scores.

In [2]:
#Load the cleaned & outcome-imputed data
train_cleaned=pd.read_csv('train_cleaned_outcome_imputed.csv')

In [3]:
#Create an initial list of predictor and outcome columns

predictors = train_cleaned.columns.tolist()
if 'id' in predictors:
    predictors.remove('id')
if 'sii' in predictors:
    predictors.remove('sii')
predictors = [x for x in predictors if 'PCIAT' not in x]
predictors = [x for x in predictors if 'Season' not in x]

**Constructing a Random Forest for Feature Identification**

In [37]:
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import FunctionTransformer


pipe_mice = Pipeline([('mice_impute', Custom_MICE_Imputer()),
                    ('add_zones', FunctionTransformer(zone_encoder)),
                    ('rf', RandomForestRegressor(n_estimators = 300, max_features = 'sqrt', max_depth = 5, random_state = 216))])

pipe_mice.fit(train_cleaned[predictors],train_cleaned['PCIAT-PCIAT_Total'])

train_pred_mice = pipe_mice.predict(train_cleaned[predictors])

#Get feature importance from the rf inside pipe
score_mice_df = pd.DataFrame({'feature':train_cleaned[predictors].columns,
                            'importance_score': pipe_mice.named_steps['rf'].feature_importances_})

score_mice_df.sort_values('importance_score',ascending=False)


feature  importance_score
0                          Basic_Demos-Age          0.145000
4                          Physical-Height          0.133568
24  PreInt_EduHx-computerinternet_hoursday          0.090154
18                             BIA-BIA_FFM          0.078891
23                       SDS-SDS_Total_Raw          0.077490
26            ENMO_Avg_Active_Days_MVPA110          0.076964
5                          Physical-Weight          0.074631
11                              FGC-FGC_CU          0.037543
19                            BIA-BIA_FFMI          0.035215
21                             BIA-BIA_Fat          0.032714
3                             Physical-BMI          0.025866
13                              FGC-FGC_PU          0.016861
28                              FGC-FGC_SR          0.016561
6             Physical-Waist_Circumference          0.014849
2                          CGAS-CGAS_Score          0.013739
30                               PAQ_Total          0.013214
10             Fitness_Endurance-Max_Stage          0.013082
32        Fitness_Endurance_Total_Time_Sec          0.011283
25            ENMO_Avg_Active_Days_MVPA192          0.010839
15                              FGC-FGC_TL          0.010569
20                             BIA-BIA_FMI          0.010162
9                     Physical-Systolic_BP          0.010094
27             Positive_Anglez_Active_Days          0.009720
17              BIA-BIA_Activity_Level_num          0.008525
22                       BIA-BIA_Frame_num          0.008482
8                       Physical-HeartRate          0.008466
7                    Physical-Diastolic_BP          0.007252
1                          Basic_Demos-Sex          0.004878
16                         FGC-FGC_TL_Zone          0.001198
29                         FGC-FGC_SR_Zone          0.000819
12                         FGC-FGC_CU_Zone          0.000729
14                         FGC-FGC_PU_Zone          0.000640
31                                PAQ_Zone          0.000000

In [4]:
keyfeatures = ['Basic_Demos-Age',
 'Physical-Height',
 'PreInt_EduHx-computerinternet_hoursday',
 'BIA-BIA_FFM',
 'SDS-SDS_Total_Raw',
 'Physical-Weight',
 'ENMO_Avg_Active_Days_MVPA110',
 'FGC-FGC_CU']

**Trying a Linear Model**

In this section, I'll make a linear model with a single predictor (hours spent on the internet)

Note: Column selector documented here: https://stackoverflow.com/questions/62416223/how-to-select-only-few-columns-in-scikit-learn-column-selector-pipeline

Note: custom loss functions for linear models are documented here: https://alexmiller.phd/posts/linear-model-custom-loss-function-regularization-python/

In [5]:
# First I'll see if I can get a pipe set up to do prediction on a split
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import FunctionTransformer


train_tt, train_ho = train_test_split(train_cleaned, test_size=0.2)

slr = Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('selector', ColumnTransformer([('selector', 'passthrough', ['PreInt_EduHx-computerinternet_hoursday'])], remainder="drop")),
                ('linear', LinearRegression())])

slr.fit(train_tt[predictors], train_tt['PCIAT-PCIAT_Total'])
mean_squared_error(train_ho['PCIAT-PCIAT_Total'], slr.predict(train_ho))

np.float64(317.8123259016001)

**An Ordinal (Sequential Binary) Classifier**

It looks like our attempts so far have under-predicted sii values of 2 and 3. 

We'll create a class that first predicts whether or not the sii value is 0, then continues upward...
This isn't quite the same as creating four separate binary predictors for 0, 1, 2, and 3 outcomes. We'll need to think about the code more to really know what it's doing.

I came up with this idea myself, but I wasn't the first one to do it. It was described on Medium: https://towardsdatascience.com/simple-trick-to-train-an-ordinal-regression-with-any-classifier-6911183d2a3c from an article by Frank and Hal

Also described on stackoverflow: https://stackoverflow.com/questions/57561189/multi-class-multi-label-ordinal-classification-with-sklearn

Some discussion of the proposed code that highlights some of its issues is on stackoverflow: https://stackoverflow.com/questions/66486947/how-to-use-ordinal-classifier

**Note:** I'll need to modify this so that it can take parameter values as inputs to the self.clf

In [ ]:
# This is something we should save in a separate file and then import
from sklearn.base import clone
from sklearn.metrics import accuracy_score

class OrdinalClassifier():

    def __init__(self, clf, **kwargs):
        self.clf = clf
        self.clf.set_params(**kwargs)
        self.clfs = {}

    def fit(self, X, y):
        self.unique_class = np.sort(np.unique(y))
        self.clf.set_params(**kwargs)
        if self.unique_class.shape[0] > 2:
            for i in range(self.unique_class.shape[0] - 1):
                # for each k - 1 ordinal value we fit a binary classification problem
                binary_y = (y > self.unique_class[i]).astype(np.uint8)
                clf = clone(self.clf)
                clf.fit(X, binary_y)
                #print('binary_y has been fit for', self.unique_class[i])
                self.clfs[i] = clf

    def predict_proba(self, X):
        self.clf.set_params(**kwargs)
        clfs_predict = {k: v.predict_proba(X) for k, v in self.clfs.items()}
        predicted = []
        for i, y in enumerate(self.unique_class):
            #print('encoding for i=', i)
            if i == 0:
                # V1 = 1 - Pr(y > V1)
                predicted.append(1 - clfs_predict[i][:, 1])
            elif y in clfs_predict:
                # Vi = Pr(y > Vi-1) - Pr(y > Vi)
                predicted.append(clfs_predict[i - 1][:, 1] - clfs_predict[i][:, 1])
            else:
                # Vk = Pr(y > Vk-1)
                predicted.append(clfs_predict[i - 1][:, 1])
        return np.vstack(predicted).T

    def predict(self, X):
        self.clf.set_params(**kwargs)
        return self.unique_class[np.argmax(self.predict_proba(X), axis=1)]
    
    def score(self, X, y, sample_weight=None):
        return accuracy_score(y, self.predict(X), sample_weight=sample_weight)

In [21]:
# Create classifiers to use as inputs to ordinal classifiers
knncordinal = OrdinalClassifier(KNeighborsClassifier(5))

In [19]:
from sklearn.preprocessing import StandardScaler, FunctionTransformer
from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

param_grid = {
    'knnordinal__n_neighbors=': [5, 10]
}

# Instantiate a random forest pipeline
ordinal_knn_pipe = Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('scale', StandardScaler()),
                ('knn_ordinal', OrdinalClassifier(KNeighborsClassifier()))])

grid_cv_knnc = GridSearchCV(ordinal_knn_pipe, 
                          param_grid = param_grid, 
                          #scoring = kappa_scorer,  #We don't want to use kappa for PCIAT predictions
                          cv = 5)

# We'll start by tuning on PCIAT, and can tune separately on sii
grid_cv_knnc.fit(train_cleaned[predictors], train_cleaned['PCIAT-PCIAT_Total'])

ValueError: Invalid parameter 'knnordinal' for estimator Pipeline(steps=[('mice_impute', Custom_MICE_Imputer()),
                ('add_zones',
                 FunctionTransformer(func=<function zone_encoder at 0x159c7fba0>)),
                ('scale', StandardScaler()),
                ('knn_ordinal',
                 <__main__.OrdinalClassifier object at 0x159d2a750>)]). Valid parameters are: ['memory', 'steps', 'verbose'].

**Comparing Models for Out-of-the-Box Performance**

**Part 1: Setting Up Models**

In the sections below, we'll set up a collection of un-tuned models and use them to predict sii scores. 

This first section instantiates the various models inside a dictionary that we'll use in the loop.

In [10]:
from sklearn.preprocessing import StandardScaler, FunctionTransformer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.svm import SVR, SVC
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor, GradientBoostingRegressor, RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier
from xgboost import XGBRegressor, XGBClassifier

# Create classifiers to use as inputs to ordinal classifiers

# Note that the logistic regression is failing to converge.
# This can be addressed - see https://stackoverflow.com/questions/62658215/convergencewarning-lbfgs-failed-to-converge-status-1-stop-total-no-of-iter
# Here we're increasing max_iter from its default and also adding a standard scaler into its pipeline
# We could also adjust the solver, ad described here: https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
logisticc = LogisticRegression(max_iter=1000)

knnc=KNeighborsClassifier(10)
svc = SVC()
rfc = RandomForestClassifier()
# Note that the default for adaboost is the SAMME.R algorithm, but this will be deprecated in future releases. Switching to SAMME
adac = AdaBoostClassifier(algorithm='SAMME')
gradc = GradientBoostingClassifier()
xgbc = XGBClassifier()


# List the various models we'll try to identify their "out of the box" performance
models = {
'slr_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('selector', ColumnTransformer([('selector', 'passthrough', ['PreInt_EduHx-computerinternet_hoursday'])], remainder="drop")),
                ('linear', LinearRegression())]),

'mlr_key_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('selector', ColumnTransformer([('selector', 'passthrough', keyfeatures)], remainder="drop")),
                ('linear', LinearRegression())]),

'mlr_all_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('linear', LinearRegression())]),

'knn_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('scale', StandardScaler()),
                ('knn', KNeighborsRegressor(10))]),

'svr_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('rf', SVR())]),

'rf_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('rf', RandomForestRegressor())]),

'ada_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('ada', AdaBoostRegressor())]),

'grad_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('grad', GradientBoostingRegressor())]),

'xgb_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('xgb', XGBRegressor())]),

'ordinal_logistic_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('scale', StandardScaler()),
                ('logistic_oc', OrdinalClassifier(logisticc))]),

'ordinal_knn_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('scale', StandardScaler()),
                ('knn_ordinal', OrdinalClassifier(knnc))]),

# Note that SVC doesn't have a predict_proba method. This can be manually added later in a custom classifier; removing for now to facilitate completion...
#'ordinal_svc_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
#                ('add_zones', FunctionTransformer(zone_encoder)),
#                ('svc_ordinal', OrdinalClassifier(svc))]),

'ordinal_rf_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('rf_ordinal', OrdinalClassifier(rfc))]),

'ordinal_ada_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('ada_ordinal', OrdinalClassifier(adac))]),

'ordinal_grad_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('grad_ordinal', OrdinalClassifier(gradc))]),

'ordinal_xgb_pipe' : Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('xgb_ordinal', OrdinalClassifier(xgbc))]),
}



**Comparing Models for Out-of-the-Box Performance**

**Part 2: Running Up Models**

The second section runs the models through a 5-fold split to compute kappa values.

We can predict sii scores in two ways:
1. Predict the PCIAT_Total score and then compute sii values
2. Predict the sii score directly

We can also modify some of these computations by adjusting the bins for computing sii values and by adjusting the computed sii scores manually.

One issue here is the extremely small number of sii=3 values in our data set. These are precisely the values we're most interested in predicting, since they most strongly indicate problematic internet use.

A suggestion from stackoverflow is to over-sample the sii=3 instances; we've implemented this over-sampling inside the k-fold split to try to avoid data leakage:
https://stackoverflow.com/questions/39512140/how-to-deal-with-this-unbalanced-class-skewed-data-set

Note: Column selector documented here: https://stackoverflow.com/questions/62416223/how-to-select-only-few-columns-in-scikit-learn-column-selector-pipeline

In [11]:
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import cohen_kappa_score

# Set up a list of the models and methods to organize the computation of means in the kfold split
modellist = []
for pipeline_name, pipeline_obj in models.items():
    modellist.append(pipeline_name)

methodlist = ['Compute SII from PCIAT (Standard Bins) (kappa)', 
              'Compute SII from PCIAT (Modified Bins) (kappa)',
                'Predict SII (rounded) (kappa)',
                'Predict SII (+0.1 rounded) (kappa)']

num_splits = 5


# Create an array with len(modellist) rows and len(methodlist) columns
output = np.zeros((len(methodlist), len(modellist), num_splits))

#Make a StratifiedKFold object stratified by the variable sii
# This is necessary due to the small number of sii=3 values
kfold = StratifiedKFold(n_splits=num_splits, shuffle=True)

## i will count the split number 
i = 0
for train_index, test_index in kfold.split(train_cleaned, train_cleaned['sii']):
#for train_index, test_index in kfold.split(train_cleaned):
    train_tt = train_cleaned.iloc[train_index]
    train_ho = train_cleaned.iloc[test_index]

    # The number of sii=3 values is so small, we'll try to boost prediction 
    # performance by duplicating the rows with this value
    train_tt_sii3=train_tt[train_tt['sii']==3]
    train_tt_sii3=pd.concat([train_tt_sii3]*4, ignore_index=True)
    train_tt=pd.concat([train_tt,train_tt_sii3], ignore_index=True)
    train_tt.reset_index(drop=True, inplace=True)

    # j will enumerate the model
    j=0

    for pipeline_name, pipeline_obj in models.items():
        # The ordinal predictors can't predict PCIAT scores, so we'll leave them out of the first round of computations
        if 'ordinal' in pipeline_name:
            kappa_sii_comp = 0
            kappa_sii_comp_mod = 0
        else:
            # Fit and make predictions of PCIAT_Total
            pipeline_obj.fit(train_tt[predictors], train_tt['PCIAT-PCIAT_Total'])
            pred = pipeline_obj.predict(train_ho[predictors])

            # Compute sii based on PCIAT and compute mse
            bins = [0, 30, 49,79,100]
            pred_bin = np.digitize(pred, bins)-1

            # Try a slightly different set of bins suggested by the "tuning" below
            bins_mod = [0, 27, 43, 71, 100]
            pred_bin_mod = np.digitize(pred, bins_mod)-1

            # Compute kappa values
            kappa_sii_comp = cohen_kappa_score(train_ho['sii'], pred_bin, weights='quadratic')
            kappa_sii_comp_mod = cohen_kappa_score(train_ho['sii'], pred_bin_mod, weights='quadratic')
        
        # Store the kappa values in the output array
        output[0,j,i] = kappa_sii_comp
        output[1,j,i] = kappa_sii_comp_mod
        j=j+1

    j=0
    for pipeline_name, pipeline_obj in models.items():
        # Fit and make predictions of sii
        pipeline_obj.fit(train_tt[predictors], train_tt['sii'])
        pred = pipeline_obj.predict(train_ho[predictors])

        # Try two different ways of rounding the predictions
        pred_round = np.round(pred)
        pred_roundmod = np.round(pred+0.1)

        # Compute and record the kappa values
        kappa_sii_round = cohen_kappa_score(train_ho['sii'], pred_round, weights='quadratic')
        kappa_sii_roundmod = cohen_kappa_score(train_ho['sii'], pred_roundmod, weights='quadratic')

        output[2,j,i] = kappa_sii_round
        output[3,j,i] = kappa_sii_roundmod
        j=j+1
    i=i+1

# Create a new array by computing the average of the values in output along the third axis
output_avg = np.mean(output, axis=2)

In [12]:
# create a data frame from output using modellist as the names of the columns and methodlist as the names of the rows
output_df = pd.DataFrame(output_avg, columns=modellist, index=methodlist)

output_df

slr_pipe  mlr_key_pipe  \
Compute SII from PCIAT (Standard Bins) (kappa)  0.334197      0.430573   
Compute SII from PCIAT (Modified Bins) (kappa)  0.347236      0.457703   
Predict SII (rounded) (kappa)                   0.280749      0.412886   
Predict SII (+0.1 rounded) (kappa)              0.335539      0.389060   

                                                mlr_all_pipe  knn_pipe  \
Compute SII from PCIAT (Standard Bins) (kappa)      0.407957  0.323621   
Compute SII from PCIAT (Modified Bins) (kappa)      0.445361  0.365642   
Predict SII (rounded) (kappa)                       0.409423  0.337445   
Predict SII (+0.1 rounded) (kappa)                  0.393304  0.334586   

                                                svr_pipe   rf_pipe  ada_pipe  \
Compute SII from PCIAT (Standard Bins) (kappa)  0.238829  0.375827  0.379687   
Compute SII from PCIAT (Modified Bins) (kappa)  0.283494  0.412270  0.412190   
Predict SII (rounded) (kappa)                   0.340159  0.346953  0.174407   
Predict SII (+0.1 rounded) (kappa)              0.346259  0.333777  0.192533   

                                                grad_pipe  xgb_pipe  \
Compute SII from PCIAT (Standard Bins) (kappa)   0.413300  0.327913   
Compute SII from PCIAT (Modified Bins) (kappa)   0.440287  0.353803   
Predict SII (rounded) (kappa)                    0.394897  0.373794   
Predict SII (+0.1 rounded) (kappa)               0.392137  0.355487   

                                                ordinal_logistic_pipe  \
Compute SII from PCIAT (Standard Bins) (kappa)                0.00000   
Compute SII from PCIAT (Modified Bins) (kappa)                0.00000   
Predict SII (rounded) (kappa)                                 0.37926   
Predict SII (+0.1 rounded) (kappa)                            0.37926   

                                                ordinal_knn_pipe  \
Compute SII from PCIAT (Standard Bins) (kappa)          0.000000   
Compute SII from PCIAT (Modified Bins) (kappa)          0.000000   
Predict SII (rounded) (kappa)                           0.282218   
Predict SII (+0.1 rounded) (kappa)                      0.282218   

                                                ordinal_rf_pipe  \
Compute SII from PCIAT (Standard Bins) (kappa)         0.000000   
Compute SII from PCIAT (Modified Bins) (kappa)         0.000000   
Predict SII (rounded) (kappa)                          0.316501   
Predict SII (+0.1 rounded) (kappa)                     0.316501   

                                                ordinal_ada_pipe  \
Compute SII from PCIAT (Standard Bins) (kappa)          0.000000   
Compute SII from PCIAT (Modified Bins) (kappa)          0.000000   
Predict SII (rounded) (kappa)                           0.339228   
Predict SII (+0.1 rounded) (kappa)                      0.339228   

                                                ordinal_grad_pipe  \
Compute SII from PCIAT (Standard Bins) (kappa)           0.000000   
Compute SII from PCIAT (Modified Bins) (kappa)           0.000000   
Predict SII (rounded) (kappa)                            0.420209   
Predict SII (+0.1 rounded) (kappa)                       0.420209   

                                                ordinal_xgb_pipe  
Compute SII from PCIAT (Standard Bins) (kappa)           0.00000  
Compute SII from PCIAT (Modified Bins) (kappa)           0.00000  
Predict SII (rounded) (kappa)                            0.36456  
Predict SII (+0.1 rounded) (kappa)                       0.36456

In [60]:
# Export output_df to a csv
output_df.to_csv('output_df.csv')

**Results of OOtB Investigation**

* The multiple regression - both using all predictors and using just the "key" features had among the largest kappa values.
* Gradient Boosting - both as a regressor and inside an ordinal classifier - was at or near the top of kappa values.
* Logistic regression inside an ordinal classifier did well
* Random forest performed decently for predicting PCIAT scores and then computing sii values
* XGBoost performed decently for predicting sii values both via regression and inside an ordinal classifier.

In general, adjusting the predicted sii values decreased kappa.

Predicting PCIAT and then computing sii benefitted from adjusting the cutpoints.

In the sections below, we'll "tune" the cutpoints and then try tuning a few of the models and re-testing the loop.

**Tuning the Bins**

We noticed that our model struggled to predict higher output values.

When we adjusted the values for converting PCIAT scores to sii scores, we noticed an improvement in prediction when we lowered the cutpoints.

We sought to "tune" these cutpoints. However, we need to be mindful of overfitting. 

We'll look at the combination of cutpoints that maximize kappa, and then from the top 20 (or so) select the cutpoints that are closest to the original ones (as measured by euclidean distance)

In [55]:
# We'll use the multiple linear regression with "key" features to test the cutpoints
# This model performed as good or better than most of the other (untuned) models
# Since it is quick to run, it should be a decent choice for doing this tuning

# Start by setting up the MLR pipeline
mlr_key_pipe=Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('selector', ColumnTransformer([('selector', 'passthrough', keyfeatures)], remainder="drop")),
                ('linear', LinearRegression())])

# Set the number of k-fold splits
num_splits = 5

# Set the number of different cutpoints to try
num_bincuts=15

# Create an array with len(modellist) rows and len(methodlist) columns
output = np.zeros((num_bincuts, num_bincuts,num_bincuts, num_splits))

#Make a KFold object, stratified on sii=3 values
kfold= StratifiedKFold(n_splits=num_splits, shuffle=True)

## i will count the split number 
i = 0

for train_index, test_index in kfold.split(train_cleaned, train_cleaned['sii']):
    train_tt = train_cleaned.iloc[train_index]
    train_ho = train_cleaned.iloc[test_index]

    # As above, we're going to "boost" the number of observations with sii=3 to try to improve predictive performance
    train_tt_sii3=train_tt[train_tt['sii']==3]
    train_tt_sii3=pd.concat([train_tt_sii3]*4, ignore_index=True)
    train_tt=pd.concat([train_tt,train_tt_sii3], ignore_index=True)
    train_tt.reset_index(drop=True, inplace=True)
    
    # Fit the pipe and make predictions
    mlr_key_pipe.fit(train_tt[predictors], train_tt['PCIAT-PCIAT_Total'])
    pred = mlr_key_pipe.predict(train_ho[predictors])

    # Iterate through values of the three cutpoints    
    for r in range(num_bincuts):
        for s in range(num_bincuts):
            for t in range(num_bincuts):
                bins = [0, 30-r, 49-s,79-t,100]
                pred_bin_mod = np.digitize(pred, bins)-1
                # Compute kappa for the binned predictions
                kappa_sii_comp_mod = cohen_kappa_score(train_ho['sii'], pred_bin_mod, weights='quadratic')
                output[r,s,t,i]=kappa_sii_comp_mod
                
    i=i+1

# Create a new array by computing the average of the values in output along the third axis
output_avg = np.mean(output, axis=3)

Next, we'll examine the output of the tuning

In [57]:
# Flatten the array and sort the indices in descending order
sorted_indices_flat = np.argsort(output_avg.ravel())[::-1]

#Decide how many top values you want to look at. 
n=20

# Get the flat indices of the top two values
top_flat_indices = sorted_indices_flat[:n]

# Convert the flat indices to 3D indices
top_indices = [np.unravel_index(idx, output_avg.shape) for idx in top_flat_indices]

# Retrieve the top two values
top_values = output_avg.ravel()[top_flat_indices]

print("Top values:", top_values)
print("Locations of the top values:", top_indices)

Top values: [0.46371448 0.46365268 0.46320111 0.46320111 0.46320111 0.46320111
 0.46296255 0.46289302 0.4627613  0.46267516 0.46244046 0.46244046
 0.46244046 0.46244046 0.46227087 0.46222172 0.46222172 0.46222172
 0.46222172 0.4621337 ]
Locations of the top values: [(np.int64(3), np.int64(8), np.int64(13)), (np.int64(3), np.int64(8), np.int64(12)), (np.int64(3), np.int64(8), np.int64(9)), (np.int64(3), np.int64(8), np.int64(11)), (np.int64(3), np.int64(8), np.int64(8)), (np.int64(3), np.int64(8), np.int64(10)), (np.int64(3), np.int64(7), np.int64(13)), (np.int64(3), np.int64(7), np.int64(12)), (np.int64(3), np.int64(6), np.int64(13)), (np.int64(3), np.int64(6), np.int64(12)), (np.int64(3), np.int64(7), np.int64(11)), (np.int64(3), np.int64(7), np.int64(9)), (np.int64(3), np.int64(7), np.int64(8)), (np.int64(3), np.int64(7), np.int64(10)), (np.int64(3), np.int64(8), np.int64(14)), (np.int64(3), np.int64(6), np.int64(10)), (np.int64(3), np.int64(6), np.int64(9)), (np.int64(3), np.int64(6

**Results of Cutpoint Tuning**

Of the top 20 cutpoints, the combination that minimizes the Euclidean distance between the original and new cutpoints is:
[0, 30-3, 49-6,79-8,100] = [0, 27, 43, 71, 100]

We'll circle back around and try this in the loop above.

**Hyperparameter Tuning**

The models above were run "out of the box." In the next sections, we'll try to tune a few of them (random forest, gradient boost, logistic regression inside an ordinal classifier, and, maybe, xgboost) to see how much we can improve their performance.

**Tuning the Random Forest Regressor**

In this section, we'll tune the random forest regressor.

Note that using a pipeline with gridcvsearch requires slightly different naming conventions for the parameter grid: 
https://stackoverflow.com/questions/34889110/random-forest-with-gridsearchcv-error-on-param-grid

Additional suggestions for only applying preprocessing before doing the gridsearchcv (not sure if we need this):
https://stackoverflow.com/questions/43366561/use-sklearns-gridsearchcv-with-a-pipeline-preprocessing-just-once

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline

param_grid = {
    'rf__n_estimators': [90, 95, 100, 105, 110],
    'rf__max_depth': [6],
    'rf__min_samples_split': [4]
}

# Instantiate a random forest pipeline
rf_pipe = Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('rf', RandomForestRegressor())])

grid_cv_rf = GridSearchCV(rf_pipe, 
                          param_grid = param_grid, 
                          #scoring = kappa_scorer,  #We don't want to use kappa for PCIAT predictions
                          cv = 5)

# We'll start by tuning on PCIAT, and can tune separately on sii
grid_cv_rf.fit(train_cleaned[predictors], train_cleaned['PCIAT-PCIAT_Total'])


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('mice_impute', Custom_MICE_Imputer()),
                                       ('add_zones',
                                        FunctionTransformer(func=<function zone_encoder at 0x152247b00>)),
                                       ('rf', RandomForestRegressor())]),
             param_grid={'rf__max_depth': [6], 'rf__min_samples_split': [4],
                         'rf__n_estimators': [90, 95, 100, 105, 110]})

In [ ]:
# Report the results
print('Best parameters:',grid_cv_rf.best_params_)
print('Best score:',grid_cv_rf.best_score_)

{'rf__max_depth': 6, 'rf__min_samples_split': 4, 'rf__n_estimators': 95}

**Results from GridSearch for Random Forest Regressor**

The first run took 30 minutes using the following parameter grid:
* 'rf__n_estimators': [50, 100, 200, 500],
* 'rf__max_depth': range(1,10),
* 'rf__min_samples_split': [2, 4, 8]

The best parameters, achieving a score of 0.28379 were:
* max_depth: 6
* min_samples_split: 4
* n_estimators: 100

The second run took 4 minutes, using a narrower ranges of parameters:
* 'rf__n_estimators': [75, 100, 150],
* 'rf__max_depth': range(4,7),
* 'rf__min_samples_split': [3, 4, 5]

The best parameters, achieving a score of 0.28396 were:
* max_depth: 6
* min_samples_split: 4
* n_estimators: 100

We will try one more search, further narrowing the ranges:
* 'rf__n_estimators': [90, 95, 100, 105, 110],
* 'rf__max_depth': [6],
* 'rf__min_samples_split': [4]

The best parameters, achieving a score of 0.2812376 were:
* n_estimators: 105

We noticed that this score was lower than the previous run, so we tried re-running. This time, the best parameters, achieving a score of 0.280982 were:
* n_estimators: 110

A third run yielded a score of 0.2808725 with
* n_estimators: 95

So it seems like the original set of best parameters (6, 4, 100) will likely 

In [ ]:
grid_cv_rf.best_estimator_.predict(train_ho[predictors])

**Tuning Random Forest for Predicting sii**

In the section above, we trained a random forest regressor to predict PCIAT scores.

Next, we'll train it for sii scores. 

GridSearchCV doesn't include kappa as a built-in metric. In addition, RF will use regression to predict sii scores, and these will need to be rounded.

So we'll need to make a custom scorer to evaluate the models.

In [51]:
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.metrics import cohen_kappa_score, make_scorer

# Create a custom scorer that first rounds the values of sii and then applies a cohen_kappa_score
def sii_kappa(y_true, y_pred):
    y_pred_round = np.round(y_pred)
    return cohen_kappa_score(y_true, y_pred_round, weights='quadratic')

kappa_scorer = make_scorer(sii_kappa)

param_grid = {
    'rf__n_estimators': [50, 75, 100],
    'rf__max_depth': range(5, 7),
    'rf__min_samples_split': [8, 10, 12]
}

# Instantiate a random forest pipeline
rf_pipe_sii = Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('rf', RandomForestRegressor())])

grid_cv_rf_sii = GridSearchCV(rf_pipe_sii, 
                          param_grid = param_grid, 
                          scoring = kappa_scorer, 
                          cv = 5)

# We'll start by tuning on PCIAT, and can tune separately on sii
grid_cv_rf_sii.fit(train_cleaned[predictors], train_cleaned['sii'])


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('mice_impute', Custom_MICE_Imputer()),
                                       ('add_zones',
                                        FunctionTransformer(func=<function zone_encoder at 0x152247b00>)),
                                       ('rf', RandomForestRegressor())]),
             param_grid={'rf__max_depth': range(5, 7),
                         'rf__min_samples_split': [8, 10, 12],
                         'rf__n_estimators': [50, 75, 100]},
             scoring=make_scorer(sii_kappa, response_method='predict'))

In [52]:
# Report the results
print('Best parameters:',grid_cv_rf_sii.best_params_)
print('Best score:',grid_cv_rf_sii.best_score_)

Best parameters: {'rf__max_depth': 6, 'rf__min_samples_split': 8, 'rf__n_estimators': 50}
Best score: 0.3586601405548822


**Results from GridSearch for Random Forest Regressor so sii**

The first run took 14 minutes using the following parameter grid:
* 'rf__n_estimators': [50, 100, 200],
* 'rf__max_depth': range(1,11),
* 'rf__min_samples_split': [2, 4, 8]

The best parameters, achieving a score of 0.36574 were:
* max_depth: 8
* min_samples_split: 8
* n_estimators: 50

The second run took 4 minutes, using a narrower ranges of parameters:
* 'rf__n_estimators': [25, 50, 75],
* 'rf__max_depth': range(5,7),
* 'rf__min_samples_split': [6, 8, 10]

The best parameters, achieving a score of 0.364877 were:
* max_depth: 6
* min_samples_split: 10
* n_estimators: 75

We will try one more search, further narrowing the ranges:
* 'rf__n_estimators': [50, 75, 100],
* 'rf__max_depth': range(5, 7),
* 'rf__min_samples_split': [8, 10, 12]

The best parameters, achieving a score of 0.3576324 were:
* max_depth: 6
* min_samples_split: 10
* n_estimators: 50

Re-running, The best parameters, achieving a score of 0.35866 were:
* max_depth: 6
* min_samples_split: 8
* n_estimators: 50

So it seems like the set of best parameters (6, 8, 50) will likely be ideal

**Tuning Gradient Boosting Regressor for PCIAT**

There are lots of parameters to consider here.

Starting to format the parameter grid using suggestions from StackOverflow: https://stackoverflow.com/questions/49500313/tune-parameters-in-gradient-boosting-reggression-with-cross-validation-sklearn

Thinking about parameters based on other articles:
* https://medium.com/all-things-ai/in-depth-parameter-tuning-for-gradient-boosting-3363992e9bae
* https://www.analyticsvidhya.com/blog/2016/02/complete-guide-parameter-tuning-gradient-boosting-gbm-python/

Note that we have a total of ~2500 observations with PCIAT scores in the entire dataset, a very small (32) number of scores with sii=3 and a relatively small number of scores with sii=2

Thus, we can limit the depth to 10 and also include relatively low values for min_samples, but should be very cautious about overfitting with small leaves and deep trees.

In [ ]:
param_grid = {
    'grad__learning_rate': [0.5, 0.1, 0.05],
    'grad__max_depth': np.linspace(3, 6, 4, endpoint=True).astype(int),
    'grad__max_features': np.linspace(1,30, 5).astype(int),
    'grad__min_samples_leaf': [1, 3, 5],
    'grad__min_samples_split': np.linspace(2, 10, 3, endpoint=True).astype(int),
    'grad__n_estimators': [10, 50, 100, 200]
}

# Going to boost the number of sii=3 values in the data set
## Note - this is potentially problematic. We should only be boosting the sii=3 values in the tt set but not the holdout set
## So we should probably try re-running this without the boost (or do the grid search manually...)

# As above, we're going to "boost" the number of observations with sii=3 to try to improve predictive performance
#train_cleaned_sii3=train_cleaned[train_cleaned['sii']==3]
#train_cleaned_sii3=pd.concat([train_cleaned_sii3]*4, ignore_index=True)
#train_tt=pd.concat([train_cleaned,train_cleaned_sii3], ignore_index=True)
#train_tt.reset_index(drop=True, inplace=True)

# Instantiate a gradient boosting regressor pipeline
grad_pipe = Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('grad', GradientBoostingRegressor())])

grid_cv_grad = GridSearchCV(grad_pipe, 
                          param_grid = param_grid, 
                          #scoring = kappa_scorer,  #We don't want to use kappa for PCIAT predictions
                          cv = 5)

# We'll start by tuning on PCIAT, and can tune separately on sii
grid_cv_grad.fit(train_cleaned[predictors], train_cleaned['PCIAT-PCIAT_Total'])


/opt/anaconda3/envs/erdos_fall_2024/lib/python3.12/site-packages/sklearn/model_selection/_validation.py:540: FitFailedWarning: 
4500 fits failed out of a total of 13500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
4500 fits failed with the following error:
Traceback (most recent call last):
  File "/opt/anaconda3/envs/erdos_fall_2024/lib/python3.12/site-packages/sklearn/model_selection/_validation.py", line 888, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/opt/anaconda3/envs/erdos_fall_2024/lib/python3.12/site-packages/sklearn/base.py", line 1473, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/erdos_fall_2024/li

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('mice_impute', Custom_MICE_Imputer()),
                                       ('add_zones',
                                        FunctionTransformer(func=<function zone_encoder at 0x152247b00>)),
                                       ('grad', GradientBoostingRegressor())]),
             param_grid={'grad__learning_rate': [0.5, 0.1, 0.05],
                         'grad__max_depth': array([3, 4, 5, 6]),
                         'grad__max_features': array([ 1,  8, 15, 22, 30]),
                         'grad__min_samples_leaf': [1, 3, 5],
                         'grad__min_samples_split': array([1, 5, 9]),
                         'grad__n_estimators': [5, 10, 50, 100, 200]})

In [68]:
# Report the results
print('Best parameters:',grid_cv_grad.best_params_)
print('Best score:',grid_cv_grad.best_score_)

Best parameters: {'grad__learning_rate': 0.05, 'grad__max_depth': np.int64(6), 'grad__max_features': np.int64(8), 'grad__min_samples_leaf': 1, 'grad__min_samples_split': np.int64(5), 'grad__n_estimators': 200}
Best score: 0.3763749996902974


*Results of GradBoost Tuning for Predicting PCIAT*

The first run took 280 minutes to run(!!!)

Unfortunately, 4500 fits failed out of a total of 13500. This is because the 'min_samples_split' parameter of GradientBoostingRegressor must be an int in the range [2, inf) and, foolishly, I had np.linspace(1, 9, 3) for this parameter.

Best parameters, yielding a score of 0.3763749996902974:
* 'grad__learning_rate': 0.05
* 'grad__max_depth': np.int64(6)
* 'grad__max_features': np.int64(8)
* 'grad__min_samples_leaf': 1
* 'grad__min_samples_split': np.int64(5)
* 'grad__n_estimators': 200

There is an issue in the first run: I had "boosted" the number of sii=3 observations. But, the way the cv works, (1) I don't think it does a stratified kfold, and (2) it has the additional sii=3 values in both the tt and holdout sets. This might artificially inflate the prediction scores.

Another issue is that the min_samples_leaf was equal to 1. This seems like it might be the result of overfitting.

I'll try running this again, but making some adjustments to the parameters. Before that, I'll set up additional tuning and run everything overnight.

**Tuning the Gradiant Boosting Regressor for sii**

In [ ]:
param_grid = {
    'grad__learning_rate': [0.5, 0.1, 0.05],
    'grad__max_depth': np.linspace(3, 6, 4, endpoint=True).astype(int),
    'grad__max_features': np.linspace(1,30, 5).astype(int),
    'grad__min_samples_leaf': [1, 3, 5],
    'grad__min_samples_split': np.linspace(2, 10, 3, endpoint=True).astype(int),
    'grad__n_estimators': [10, 50, 100, 200]
}

# Create a custom scorer that first rounds the values of sii and then applies a cohen_kappa_score
def sii_kappa(y_true, y_pred):
    y_pred_round = np.round(y_pred)
    return cohen_kappa_score(y_true, y_pred_round, weights='quadratic')

kappa_scorer = make_scorer(sii_kappa)

# Instantiate a gradient boosting regressor pipeline
grad_pipe_sii = Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('grad', GradientBoostingRegressor())])

grid_cv_grad_sii = GridSearchCV(grad_pipe, 
                          param_grid = param_grid, 
                          scoring = kappa_scorer,=
                          cv = 5)

grid_cv_grad_sii.fit(train_cleaned[predictors], train_cleaned['sii'])

In [ ]:
# Report the results
print('Best parameters:',grid_cv_grad_sii.best_params_)
print('Best score:',grid_cv_grad_sii.best_score_)

**Tuning the Logistic Regression inside an Ordinal Predictor**

First I'll try to do this using gridsearchcv. But that seems to not be working very well....

In [ ]:
param_grid = {
    #Insert logistic regression parameters here
}

# Create a custom scorer that first rounds the values of sii and then applies a cohen_kappa_score
def sii_kappa(y_true, y_pred):
    y_pred_round = np.round(y_pred)
    return cohen_kappa_score(y_true, y_pred_round, weights='quadratic')

kappa_scorer = make_scorer(sii_kappa)

# Instantiate an ordinal classifier pipe
ordinal_logistic_pipe = Pipeline([('mice_impute', Custom_MICE_Imputer()),
                ('add_zones', FunctionTransformer(zone_encoder)),
                ('scale', StandardScaler()),
                ('logistic_oc', OrdinalClassifier(LogisticRegression(max_iter=1000)))])

grid_cv_logistic = GridSearchCV(grad_pipe, 
                          param_grid = param_grid, 
                          scoring = kappa_scorer,=
                          cv = 5)

grid_cv_logistic.fit(train_cleaned[predictors], train_cleaned['sii'])

**Manually Tuning Logistic Regression Inside an Ordinal Classifier**

Going to try to do the tuning manually

In [ ]:
## import random forest classifier
from sklearn.ensemble import RandomForestClassifier

## import kfold
from sklearn.model_selection import StratifiedKFold

## import accuracy_score
from sklearn.metrics import accuracy_score

## set the number of CV folds
n_splits = 5

## Make the kfold object
kfold = StratifiedKFold(n_splits, 
                        random_state=216, 
                        shuffle=True)

## note this will take about 2 minutes to run

max_depths = range(1, 11)
n_trees = [100, 500]

rf_accs = np.zeros((n_splits, len(max_depths), len(n_trees)))



for i,(train_index, test_index) in enumerate(kfold.split(seeds_train, seeds_train.y)):
    seeds_tt = seeds_train.iloc[train_index]
    seeds_ho = seeds_train.iloc[test_index]

    for j, max_depth in enumerate(max_depths):
        for k, n_estimators in enumerate(n_trees):
            print(i,j,k)
            rf = RandomForestClassifier(n_estimators = n_estimators,
                                           max_depth = max_depth,
                                           max_samples = 0.8,
                                           random_state = 216)
                                           
            rf.fit(seeds_tt[features], seeds_tt.y)
            
            pred = rf.predict(seeds_ho[features])
            
            rf_accs[i,j,k] = accuracy_score(seeds_ho.y,  pred)

In [ ]:
max_index = np.unravel_index(np.argmax(np.mean(rf_accs, axis=0)), 
                                       np.mean(rf_accs, axis=0).shape)


print(max_depths[max_index[0]],n_trees[max_index[1]])